janusgraph-core/src/main/java/org/janusgraph/graphdb/query/index/ApproximateIndexSelectionStrategy.java

```diff
@@ -0,0 +1,83 @@
+import org.janusgraph.graphdb.database.IndexSerializer;
+import org.janusgraph.graphdb.internal.OrderList;
+import org.janusgraph.graphdb.query.condition.Condition;
+import org.janusgraph.graphdb.query.condition.MultiCondition;
+import org.janusgraph.graphdb.query.graph.JointIndexQuery;
+import org.janusgraph.graphdb.types.IndexType;
+import org.janusgraph.graphdb.types.MixedIndexType;
+import org.javatuples.Pair;
+
+/**
+ * @author Florian Grieskamp (Florian.Grieskamp@gdata.de)
+ */
+public class ApproximateIndexSelectionStrategy
+    extends AbstractIndexSelectionStrategy implements IndexSelectionStrategy {
+
+    /**
+     * Iterate over all potential indexes and compute a score based on how many clauses
+     * this index covers. The index with the highest score (as long as it covers at least one
+     * additional clause) is picked and added to the joint query for as long as such exist.
+     */
+    @Override
+    public SelectedIndexQuery selectIndices(final Set<IndexType> rawCandidates,
+                                            final MultiCondition<JanusGraphElement> conditions,
+                                            final Set<Condition> coveredClauses, OrderList orders,
+                                            IndexSerializer serializer) {
+        final JointIndexQuery jointQuery = new JointIndexQuery();
+        boolean isSorted = orders.isEmpty();
+        while (true) {
+            IndexCandidate bestCandidate = null;
+            boolean candidateSupportsSort = false;
+
+            for (final IndexType index : rawCandidates) {
+                final IndexCandidate indexCandidate =
+                    createIndexCandidate(index, conditions, serializer);
+                if (indexCandidate == null) {
+                    continue;
+                }
+
+                boolean supportsSort =
+                    orders.isEmpty() ||
+                    coveredClauses.isEmpty() && index.isMixedIndex() &&
+                        IndexSelectionStrategy.indexCoversOrder((MixedIndexType) index, orders);
+                indexCandidate.calculateScoreApproximation(coveredClauses, supportsSort);
+
+                if (!indexCandidate.isCoveredBy(coveredClauses) &&
+                    (bestCandidate == null ||
+                     indexCandidate.getScore() > bestCandidate.getScore())) {
+                    bestCandidate = indexCandidate;
+                    candidateSupportsSort = supportsSort;
+                }
+            }
+
+            if (bestCandidate != null) {
+                if (coveredClauses.isEmpty()) {
+                    isSorted = candidateSupportsSort;
+                }
+                coveredClauses.addAll(bestCandidate.getSubCover());
+                addToJointQuery(bestCandidate, jointQuery, serializer, orders);
+            } else {
+                break;
+            }
+        }
+        return new SelectedIndexQuery(jointQuery, Boolean.valueOf(isSorted));
```

Comment:
Redundant boxing. I.e. use `isSorted` directly instead of `Boolean.valueOf(isSorted)`